### ISSUE TITLE
save_as_dataset should complain when trying to save data with incompatible shapes

### ISSUE THREAD

**Issue by ngoldbaum** [yt-project/yt#1609]
### Bug report

**Bug summary**

save_as_dataset is happy to save unstructured field data with different shapes. This leads to errors when trying to actually use the data later.

**Code for reproduction**

```python
import yt
import numpy as np

density = np.random.random(128)
temperature = np.random.random(128)
cell_mass = np.random.random((128, 128))

my_data = {
    "density": density,
    "temperature": temperature,
    "cell_mass": cell_mass}
fake_ds_med = {"current_time": yt.YTQuantity(10, "Myr")}
yt.save_as_dataset(fake_ds_med, "mydata.h5", my_data)

ds = yt.load('mydata.h5')

yt.PhasePlot(ds.data, 'temperature', 'density', 'cell_mass')
```

**Actual outcome**

```
Traceback (most recent call last):
  File "testt.py", line 17, in <module>
    yt.PhasePlot(ds.data, 'temperature', 'density', 'cell_mass')
  File "/Users/goldbaum/Documents/yt-git-fixes/yt/visualization/profile_plotter.py", line 761, in __init__
    fractional=fractional)
  File "/Users/goldbaum/Documents/yt-git-fixes/yt/data_objects/profiles.py", line 1095, in create_profile
    obj.add_fields([field for field in fields])
  File "/Users/goldbaum/Documents/yt-git-fixes/yt/data_objects/profiles.py", line 127, in add_fields
    self._bin_chunk(chunk, fields, temp_storage)
  File "/Users/goldbaum/Documents/yt-git-fixes/yt/data_objects/profiles.py", line 567, in _bin_chunk
    rv = self._get_data(chunk, fields)
  File "/Users/goldbaum/Documents/yt-git-fixes/yt/data_objects/profiles.py", line 255, in _get_data
    arr[:,i] = chunk[field][filter].in_units(units)
ValueError: could not broadcast input array from shape (128,128) into shape (128)
```

**Expected outcome**

I'd expect yt to raise an error during the save_as_dataset call.

**Version Information**

  * Operating System:
  * Python Version: 3.6
  * yt version: 3.5dev
  * Other Libraries (if applicable):

**Comment by ngoldbaum** [comment]
This was motivated by this mailing list thread:

http://lists.spacepope.org/pipermail/yt-users-spacepope.org/2017-November/008948.html

@brittonsmith do you agree that we should be raising an error in these cases?

**Comment by brittonsmith** [comment]
@ngoldbaum, no I don't think we should raise an error for differently shaped data.  This is something I routinely do when I just want to save a bunch of random data arrays that I will later reload and access through `ds.data`.

I agree that the above script shouldn't work, but I don't `yt.save_as_dataset` was meant to guarantee that re-loaded data can be used for any yt operation.

**Comment by ngoldbaum** [comment]
Ah, I didn’t realize that was intended. Ok, in that case what about adding a check in create_profile so that we fail with a nicer error message?

**Comment by brittonsmith** [comment]
As a followup, I'm not sure this is exactly the same, but you can do something like the following:
```
ad = ds.all_data()
p = yt.create_profile(ad, ["density", "temperature"], "cell_mass", weight_field=None)
fn = p.save_as_dataset()

prof_ds = yt.load(fn)
# access the profile
prof_ds.profile.x
prof_ds.profile.y
prof_ds.profile["cell_mass"]
```

**Comment by brittonsmith** [comment]
Yeah, I agree with adding a check to `create_profile` so it doesn't barf somewhere in the guts.

**Comment by brittonsmith** [comment]
Resolved in PR #2015.
